### DISCUSSION TITLE
Need clarification/input on my solution in fixing blank chart when the app is initially launched

### DISCUSSION THREAD

**Discussion by LeeWeeder** [patrykandpatrick/vico#770] [Questions] [answered]
### Question

Good day, all! Firstly, I would like to thank you for this wonderful library!

So, as for the title, I was having a problem with displaying a line chart on the Home Screen. The home screen is the first screen loaded when the app is launched, however, the line chart would be empty—the data was for sure not empty but since it's from the initial launch of the app, it was empty.  When I logged the data size and the corresponding tryRunTransaction result, it returned two logs; zero and true, respectively, on the first, and 5 and false on the second. However, after navigating from the other screen and back to the Home Screen, the line chart will be now populated. 

Here is the code of the line chart:
```
val xToDateMapKey = ExtraStore.Key<Map<Int, Pair<Float, Instant>>>()

@Composable
fun LineChart(uiState: HomeUiState) {
    ElevatedCard(
        modifier = Modifier
            .padding(vertical = 16.dp, horizontal = 8.dp),
        onClick = {}
    ) {
        val modelProducer =
            remember { CartesianChartModelProducer.build {} }

        val data = uiState.fiveMostRecentLogs.reversed().associate {
            it.id to Pair(it.weight.value, it.date)
        }

        LaunchedEffect(uiState.fiveMostRecentLogs) {
            val xToDates = data.mapValues { (_, pair) ->
                LocalDate.ofInstant(pair.second, ZoneId.systemDefault()).toEpochDay()
                    .toFloat() to pair.second
            }

            modelProducer.tryRunTransaction {
                lineSeries {
                    series(
                        x = xToDates.keys, y = data.values.map { it.first }
                    )
                }
                updateExtras {
                    it[xToDateMapKey] = xToDates
                }
            }
        }

        Column(modifier = Modifier.padding(16.dp)) {
            SectionLabel(title = "Trend")
            Spacer(modifier = Modifier.height(8.dp))
            CartesianChartHost(
                chart = rememberCartesianChart(
                    rememberLineCartesianLayer(
                        lines = listOf(
                            rememberLineSpec(
                                backgroundShader = null,
                                shader = DynamicShader.color(MaterialTheme.colorScheme.primary)
                            )
                        ),
                        axisValueOverrider = remember {
                            AxisValueOverrider.adaptiveYValues(1f, true)
                        }
                    ),
                    startAxis = rememberStartAxis(),
                    bottomAxis =
                        rememberBottomAxis(
                            valueFormatter = { x, chartValues, _ ->
                                (chartValues.model.extraStore[xToDateMapKey][x.toInt()] ?: Pair(
                                    x,
                                    Instant.ofEpochSecond(LocalDate.ofEpochDay(x.toLong()).toEpochDay() * 86400)
                                )).second.getFormattedDate("MM/dd/yy", useRelativeDates = true)
                            }
                        )
                ),
                modelProducer = modelProducer
            )
        }
    }
}

```

My solution:
Wrap the series with not empty check and provide a placeholder on the else block
```
lineSeries {
    if (xToDates.isNotEmpty() && data.isNotEmpty()) {
        series(
            x = xToDates.keys, y = data.values.map { it.first }
        )
    } else {
        series(
            x = mapOf(
                1 to Pair(1f, Instant.now())
            ).keys, y = listOf(2.0)
        )
    }
}
```
Although the else block series is not empty, the series under it was not displayed on the initial launch but the data under the if block is, instead, thus fixing the line chart not showing.

I would like to ask for your input on this.

Edit:
The above solution only works when the `uiState.fiveMostRecentLogs` is not empty otherwise the app crashes on launch and throws Cannot round NaN value.

Thank you!

### Vico version(s)

2.0.0-alpha.21

### UI framework(s)

Jetpack Compose

**Comment by Gowsky** [comment, marked as answer]
Hello @LeeWeeder,
>When I logged the data size and the corresponding tryRunTransaction result, it returned two logs; zero and true, respectively, on the first, and 5 and false on the second. 

It looks like your data is initially empty and quickly updated. As mentioned in the wiki, `runTransaction` should be used for frequent data updates. However, adding empty data is incorrect and you must avoid it. See the first paragraph [here](https://github.com/patrykandpatrick/vico/discussions/734#discussioncomment-9585013) for more information. You should be running the transaction when your data is not empty. If this is the only data update, you can stick to `tryRunTransaction`, otherwise use `runTransaction`.

On a side note, you are blocking the main thread with your data processing. It all should be run inside of `LaunchedEffect` and a non-main dispatcher.
